.github/workflows/linux.yml

```diff
@@ -15,520 +10,97 @@ on:
-      - name: Validate build
-        run: sh validate.sh -j 2 -w ghc-8.8.4 -v --solver-benchmarks -s build
-      - name: Validate lib-tests
-        run: sh validate.sh -j 2 -w ghc-8.8.4 -v --solver-benchmarks -s lib-tests
-      - name: Validate lib-suite
-        run: sh validate.sh -j 2 -w ghc-8.8.4 -v --solver-benchmarks -s lib-suite
-      - name: Validate cli-tests
-        run: sh validate.sh -j 2 -w ghc-8.8.4 -v --solver-benchmarks -s cli-tests
-      - name: Validate cli-suite
-        run: sh validate.sh -j 2 -w ghc-8.8.4 -v --solver-benchmarks -s cli-suite
-  validate-8_6_5:
-    name: validate.sh ghc-8.6.5
-    runs-on: ubuntu-18.04
-    needs: validate-8_8_4
-    container:
-      image: phadej/ghc:8.6.5-bionic
+  validate:
+    name: Validate ghc-${{ matrix.ghc }}
+
+    runs-on: ubuntu-latest
+    continue-on-error: ${{ matrix.experimental }}
+
+    strategy:
+      matrix:
+        include:
+          - ghc: "9.2.1"
+            cli: true
+            experimental: true
+            flags: ""
+          - ghc: "9.0.2"
+            cli: false
+            experimental: true # FIXME for quick testing false
+            flags: "--lib-only"
+          - ghc: "8.10.7"
+            cli: true
+            experimental: true # FIXME for quick testing false
+            flags: ""
+          - ghc: "8.8.4"
+            cli: true
+            experimental: true # FIXME for quick testing false
+            flags: "--solver-benchmarks"
+          - ghc: "8.6.5"
+            cli: true
+            experimental: true # FIXME for quick testing false
+            flags: "--complete-hackage-tests"
+          - ghc: "8.4.4"
+            cli: true
+            experimental: true # FIXME for quick testing false
+            flags: ""
+          - ghc: "8.2.2"
+            cli: true
+            experimental: true # FIXME for quick testing false
+            flags: ""
+          - ghc: "8.0.2"
+            cli: false
+            experimental: true # FIXME for quick testing false
+            flags: "--lib-only"
+
     steps:
-      - name: System info
-        run: |
-          uname -a
-      - name: Set PATH
-        # https://docs.github.com/en/free-pro-team@latest/actions/reference/workflow-commands-for-github-actions#adding-a-system-path
-        run: |
-          echo "$HOME/.cabal/bin" >> $GITHUB_PATH
+
+      - uses: actions/checkout@v2
+
+      - uses: haskell/actions/setup@v1
+        id: setup-haskell
+        with:
+          ghc-version: ${{ matrix.ghc }}
+          cabal-version: latest # default, keeping for visibility
+
+      #  See the following link for a breakdown of the following step
+      #  https://github.com/haskell/actions/issues/7#issuecomment-745697160
+      - uses: actions/cache@v2
+        with:
+          path: |
+            ${{ steps.setup-haskell.outputs.cabal-store }}
+            dist-newstyle
+          key: ${{ runner.os }}-${{ matrix.ghc }}-${{ github.sha }}
```

Comment:
this cache key is too restrictive, hashing all files which can control the cache (*.cabal, *.project) would be a still to broader one, doing for the specific config files we know are being used would be beven better

Reply:
ah and the hackage index being used could be useful too

Reply:
This is a bit of a trick, the comment I left above is relevant.

Our cache is basically append only. We want GitHub to always save and restore the cache. To get it to do that we use a key that (as you notice) is unique but we also instruct it to use as a backup anything that matches up to `${{ runner.os }}-${{ matrix.ghc }}-`. GitHub will pick the most recent matching key.

If we used a same key (for example just `${{ runner.os }}-${{ matrix.ghc }}`) GitHub would not update the cache.

So the alterative is 1) either we are 100% precise (and I guess we could) 2) or we make it append only with the trick above.

See https://github.com/haskell/actions/issues/7#issuecomment-745697160 for the original discussion

Reply:
Thanks for the pointer to the coment, very interesting
Maybe i am missing something obvious but usually all prs are adding new commits so the real used cache is the fallback of the previous (or whatever) commit. It will take lot of cache space as each commit in the repo will augment it and the limit is 20 Gb, after that gh will start to evict old caches. But most of them will be identical (not sure if gh is able to optimize it underneath). If you returns to a old pr you will get a first rebuild for sure.
Otoh one commit could have very different dependencies and they change in the cabal store not only by version but with thing like ghc-options. Interleaved commits between prs will invalidate the cache for all them causing unnecessary rebuilds.


The comment talks about using the cabal.project.freeze, which is even better than hashing cabal.project and .cabal files, as it captures cli flags. I nfact it is what we are using in hls (with the hackage index as base key):

https://github.com/haskell/haskell-language-server/blob/96ea854debd92f9a54e2270b9b9a080c0ce6f3d1/.github/actions/setup-build/action.yml#L131-L142
But well cabal dependency set is relatively small and usually there are no lot simultaneous prs so maybe it can work in practice.

Reply:
forgot I said about interleaved commits, as all commits are appending library with duff hashes in the cabal store, all of them will have what they need

Reply:
Rephrasing, GitHub cache is immutable, given a key, once it's built, it is assumed to be correct, complete and "done".

On the other hand, cabal store is append only, it is its own cache if you want. Each package gets built only once (per configuration).

The trick is to make the former behave like the latter.

You raise an interesting point and I believe what you are saying is correct. If I understand you, you are saying that given the cache is not linked to the tree at all, the most recent cache might be inappropriate for the branch you are building.
Nevertheless I don't believe this is going to be a problem, we only use order of ~50MB per cache entry (link](https://github.com/haskell/cabal/runs/5218127255?check_suite_focus=true#step:4:10)). I think with 20GB we can build half hackage and keep it cached :-)

Using `cabal.project.freeze` would mean that any change in the plan will throw away the entire cache. E.g. if we change an option in `cabal-install` we will end up rebuilding `criterion` (a dependency of `cabal-benchmark`).

If anything the real issue I see with this setup is that the cache doens't get saved unless the workflow succeed. This has caused unneeded rebuilding of the dependencies, for example. Ideally, I would like the cabal store be persisted regardless of the workflow success. It's a known issue with `actions/cache`, see [here](https://github.com/actions/cache/issues/92).

Reply:
yeah with that small dep footprint you can let grow the store quite time
In hls each key matrix has 200mb and we can have ~20 combinations of os+ghc so we have to reuse cache more aggressively even having to rebuild part of deps
But even with 20 mb you will reach the limit at some point and the actual cache key prefix is totally fixed. A manual reset changing the prefix will be needed.

In hls we have pinned the hackage index in the cabal.project and using it in the cache key gives us a pretty stable dependency set. We update the index from time to time and that make generate a clean cache shared for all PRS. So changes in the freeze file doesnt suppose a full rebuild almost never.

But hls scenario is pretty different from cabal one.


I have to take a look to the cache on pr failure issue but in theory the failure can be caused precisely by introducing a bad dep in the cache or even a corruption in the store (improbable but still, it happened to me locally) making it unusable and forcing you to invalidate manually changing the prefix. It can be avoided building only dependencies in a separate job, save the cache if it is successful and then build the local packages in other job reusing the cache.

Reply:
perhaps it could be worth exploring using an external cache, it shouldn't be hard do to with S3 `aws s3 sync s3://my-cache` and back.

Reply:
i think it will not be as fast to restore and save it as the github one, located very near but maybe i am wrong (or the perf gain is not significant, to investigate)